docs/decisions/0063-vector-store-list-keys.md

```diff
@@ -0,0 +1,118 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
+contact: eavanvalkenburg
+date: 2024-12-10
+deciders: sergeymenshykh, markwallace, rbarreto, dmytrostruk, westey-m, eavanvalkenburg, bentho
+consulted: 
+informed: 
+---
+
+# Adding list keys method to VectorStoreRecordCollection interface
+
+The new [vector store design](./0050-updated-vector-store-design.md) has been released, but a few users noticed that a method to list all existing keys is missing. This decision proposes a new method to list all keys in the vector store.
+
+## Context and problem statement
+
+Since we support CRUD operations in our data stores we should also provide a way to keep a collection in sync between our code and the backend database. This is currently not easy since it would required either querying every key to see if it exists or maintaining a separate list of keys in the code. Only when you know which keys are in the database and not in your code can you decide to delete them.
+
+## Decision drivers
+1. We want to provide a way to list all keys in the vector store
+2. We want to provide this in a non-breaking way
+3. 
+
+## Considered Options
+
+- Option #1: add a new method to the VectorStoreRecordCollection interface
+- Option #2: add a new method to a new separate interface
+- Option #3: alter the behavior of the Get/GetBatch methods to allow getting all records (key=None or similar)
+
+## Option 1: add a new method to the VectorStoreRecordCollection interface
+
+We can add a new method to the VectorStoreRecordCollection interface to list all keys in the vector store. This method will return a list of keys.
+
+```csharp
+public interface IVectorStoreRecordCollection<TKey, TRecord> : IVectorizedSearch<TRecord>
+{
+    Task<TRecord?> GetAsync(TKey key, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<TRecord> GetBatchAsync(IEnumerable<TKey> keys, GetRecordOptions? options = default, CancellationToken cancellationToken = default);
+    ...
+    Task<IEnumerable<TKey>> ListKeysAsync(CancellationToken cancellationToken = default);
+}
+```
+Pros:
+- This is a simple and straightforward way to add the functionality
+Cons:
+- It is a breaking change (at least for csharp, python can do a default implementation that returns None)
+
+## Option 2: add a new method to a new separate interface
+
+We can add a separate interface to list all keys in the vector store. This method will return a list of keys.
+
+```csharp
+public interface IVectorStoreRecordCollectionListKeys<TKey, TRecord> : IVectorStoreRecordCollection<TKey, TRecord>
+{
+    Task<IEnumerable<TKey>> ListKeysAsync(CancellationToken cancellationToken = default);
+}
+```
+Pros:
+- This is a non-breaking change, each vector store record collection must choose to implement this interface.
+
+Cons:
+- This is a more complex way to add the functionality
```

Comment:
If we decide to avoid this approach, I think we should add more details here why it's complex, so it will be easier to understand in the future why we decided to go with another option.